### PR TITLE
New: Add Transmission labels support

### LINKED ISSUE #8661: Labels support for Transmission
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues

### Is your feature request related to a problem? Please describe

At the moment the Transmission client in Radarr/Sonarr doens't support labels and it uses a fake `Category` system that put the file in a subfolder.

### Describe the solution you'd like

If I remember correctly Transmission supports labels since the version 3.0 and now it is at the version 4.0.
I think that if implemented the labels system should work like the Flood client (where labels=tags), supporting `Post-Import Tags` and `Additional Tags`.

The filed is called `labels` and it's a list of strings.

### Describe alternatives you've considered

Change torrent client

### Anything else?

[Transmission's RPC specification](https://github.com/transmission/transmission/blob/0ef58c2a20019096aed0ab639f2207b8e1c1b248/docs/rpc-spec.md#32-torrent-mutator-torrent-set)

[Add labels feature](https://github.com/transmission/transmission/pull/822) PR in the Transmission repo

### PR DESCRIPTION
**Pull request by Damix48**
#### Database Migration
NO

#### Description
This add labels support to the Transmission client.
Features added:
- [x] Category labels
- [x] Additional labels like `Collection`, `Quality`, `ReleaseGroup`, etc
- [x] Post-imported labels
- [x] Backwards compatible

Regarding backwards compatibility:
- If Transmission doesn't support labels, the old `Category` setting works as intended. If labels are set they are ignored.
- If Transmission supports labels, `Category` and all the `Label` settings work together.

#### Todos
- [ ] Tests
- [ ] Check Vuze compatibility, version checking?

#### Issues Fixed or Closed by this PR

* Fixes #8661

### CODE REVIEW

File: src/NzbDrone.Core/Download/Clients/Transmission/TransmissionProxy.cs [hunk 1]
```diff
@@ -38,16 +41,16 @@ public TransmissionProxy(ICacheManager cacheManager, IHttpClient httpClient, Log
             _authSessionIDCache = cacheManager.GetCache<string>(GetType(), "authSessionID");
         }
 
-        public List<TransmissionTorrent> GetTorrents(TransmissionSettings settings)
+        public List<TransmissionTorrent> GetTorrents(bool labels, TransmissionSettings settings)
```
[the remark this answers is not in the document]
Comment (Damix48) [reply src/NzbDrone.Core/Download/Clients/Transmission/TransmissionProxy.cs:44]: So something like this:
```csharp
        private bool _supportLabels;

        public TransmissionProxy(ICacheManager cacheManager, IHttpClient httpClient, Logger logger)
        {
            _httpClient = httpClient;
            _logger = logger;

            _authSessionIDCache = cacheManager.GetCache<string>(GetType(), "authSessionID");

            _supportLabels = GetClientVersion(new TransmissionSettings()) >= new Version(3, 0);
        }

        public bool SupportLabels => _supportLabels;
```

So, `TransmissionBase` uses `_proxy.SupportLabels` when needed.

Is this a good solution?

File: src/NzbDrone.Core/Download/Clients/Transmission/TransmissionProxy.cs [hunk 2]
```diff
@@ -177,7 +210,8 @@ private TransmissionResponse GetTorrentStatus(IEnumerable<string> hashStrings, T
                 "seedRatioMode",
                 "seedIdleLimit",
                 "seedIdleMode",
-                "fileCount"
+                "fileCount",
+                labels ? "labels" : null
```
[the remark this answers is not in the document]
Comment (Damix48) [reply src/NzbDrone.Core/Download/Clients/Transmission/TransmissionProxy.cs:214]: Okay, I'll change the array into a List

File: src/NzbDrone.Core/Download/Clients/Transmission/TransmissionBase.cs [hunk 3]
```diff
@@ -228,6 +294,23 @@ protected override string AddFromTorrentFile(RemoteMovie remoteMovie, string has
             return hash;
         }
 
+        public override void MarkItemAsImported(DownloadClientItem downloadClientItem)
+        {
+            if (SupportLabels && Settings.PostImportLabels.Any())
+            {
+                try
+                {
+                    var torrent = _proxy.GetTorrents(SupportLabels, Settings).Single(t => t.HashString == downloadClientItem.DownloadId.ToLower());
```
[the remark this answers is not in the document]
Comment (Damix48) [reply src/NzbDrone.Core/Download/Clients/Transmission/TransmissionBase.cs:303]: Done

File: src/NzbDrone.Core/Download/Clients/Transmission/TransmissionBase.cs [hunk 4]
```diff
@@ -228,6 +294,23 @@ protected override string AddFromTorrentFile(RemoteMovie remoteMovie, string has
...
+                    var labels = torrent.Labels.Concat(Settings.PostImportLabels).Distinct().ToList();
```
[the remark this answers is not in the document]
Comment (Damix48) [reply src/NzbDrone.Core/Download/Clients/Transmission/TransmissionBase.cs:304]: You mean remove the labels of `Settings.Labels` from labels of the torrent, then add the `Settings.PostImportLabels`?

File: src/NzbDrone.Core/Download/Clients/Transmission/TransmissionProxy.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Damix48) [reply src/NzbDrone.Core/Download/Clients/Transmission/TransmissionProxy.cs:44]: Yes but do a request every time

File: src/NzbDrone.Core/Download/Clients/Transmission/TransmissionBase.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (Damix48) [reply src/NzbDrone.Core/Download/Clients/Transmission/TransmissionBase.cs:304]: Maybe adding a new advance option? Something like `Delete tags after imported`

[the remark this answers is not in the document]
Comment (Damix48) [reply src/NzbDrone.Core/Download/Clients/Transmission/TransmissionBase.cs:304]: I personally prefer the way Flood manage the Tags/Labels.
I think adding an advance option in client setting is a good compromise.
By default:
- The "Labels" and "AdditionalLabels" would be removed.
- The "PostImportedLabels" would be added.

However, if the advanced option is set to false:
- The "Labels" and "AdditionalLabels" would be retained.
- The "PostImportedLabels" would still be added.

Personally, I prefer this approach as it offers more flexibility in managing torrents within the Transmission GUI.

However, adding an advanced option could potentially complicate the client setup, especially for less experienced users.
No problem if this is not a good/useful or too complicated feature.

**Comment by austinarchibald** [comment]
Hi, eagerly awaiting this feature to be added to latest develop branch. What's the latest? Thanks!

**Pull request closed without merging**